src/omnicore/fees.cpp

```diff
@@ -0,0 +1,478 @@
+                for (std::set<feeCacheItem>::iterator it = sCacheHistoryItems.begin(); it != sCacheHistoryItems.end(); it++) {
+                    feeCacheItem tempItem = *it;
+                    if (tempItem.first >= block) continue; // discard this entry
+                    if (!newValue.empty()) newValue += ",";
+                    newValue += strprintf("%d:%d", tempItem.first, tempItem.second);
+                }
+                leveldb::Status status = pdb->Put(writeoptions, key, newValue);
+                assert(status.ok());
+                PrintToLog("Rolling back fee cache for property %d, new=%s [%s])\n", propertyId, newValue, status.ToString());
+            }
+        }
+    }
+}
+
+// Evaluates fee caches for the property against threshold and executes distribution if threshold met
+void COmniFeeCache::EvalCache(const uint32_t &propertyId, int block)
+{
+    if (GetCachedAmount(propertyId) >= distributionThresholds[propertyId]) {
+        DistributeCache(propertyId, block);
+    }
+}
+
+// Performs distribution of fees
+void COmniFeeCache::DistributeCache(const uint32_t &propertyId, int block)
+{
+    LOCK(cs_tally);
+
+    int64_t cachedAmount = GetCachedAmount(propertyId);
+
+    OwnerAddrType receiversSet = STO_GetReceivers("FEEDISTRIBUTION", OMNI_PROPERTY_MSC, cachedAmount);
+    uint64_t numberOfReceivers = receiversSet.size(); // there will always be addresses holding OMNI, so no need to check size>0
+    PrintToLog("Starting fee distribution for property %d to %d recipients...\n", propertyId, numberOfReceivers);
+
+    int64_t sent_so_far = 0;
+    std::set<feeHistoryItem> historyItems;
+    for (OwnerAddrType::reverse_iterator it = receiversSet.rbegin(); it != receiversSet.rend(); ++it) {
+        const std::string& address = it->second;
+        int64_t will_really_receive = it->first;
+        sent_so_far += will_really_receive;
+        if (msc_debug_fees) PrintToLog("  %s receives %d (running total %d of %d)\n", address, will_really_receive, sent_so_far, cachedAmount);
+        assert(update_tally_map(address, propertyId, will_really_receive, BALANCE));
+        feeHistoryItem recipient(address, will_really_receive);
+        historyItems.insert(recipient);
+    }
+
+    PrintToLog("Fee distribution completed, distributed %d out of %d\n", sent_so_far, cachedAmount);
+
+    // store the fee distribution
+    p_feehistory->RecordFeeDistribution(propertyId, block, sent_so_far, historyItems);
+
+    // final check to ensure the entire fee cache was distributed, then empty the cache
+    assert(sent_so_far == cachedAmount);
+    ClearCache(propertyId, block);
+}
+
+// Prunes entries over 50 blocks old from the entry for a property
+void COmniFeeCache::PruneCache(const uint32_t &propertyId, int block)
+{
+    if (msc_debug_fees) PrintToLog("Starting PruneCache for prop %d block %d...\n", propertyId, block);
+    assert(pdb);
+
+    int pruneBlock = block-50;
```

Comment:
Fees are only 50 blocks reorg-safe that's correct - I chose that value as it's the same as the rest of the state (we only keep the last 50 blocks).

If a reorg of more than 50 blocks occurs the client clears the state and reparses from scratch - that would include fees.